# プロ野球でわかる！はじめての統計学

In [1]:
import numpy as np
import pandas as pd
import requests
import json
import numpy as np

## 偏差
```
偏差 = 元のデータ - 平均値
```

## 分散
```
分散 = (Σ偏差^2) ÷ データ数
```

## 標準偏差
```
標準偏差 = sqrt(分散)
```

In [2]:
org_data = pd.read_csv('./statistics_data/2章サンプルデータ.csv')

avg_data = org_data[['選手名','打率']]

avg_data = avg_data.astype({'打率': 'float64'})

avg_data['偏差'] = avg_data['打率'] - avg_data.mean()['打率']

avg_data['偏差^2'] = avg_data['偏差'] * avg_data['偏差']

dispersion = avg_data.sum()['偏差^2'] / 185
print(dispersion)

0.0014787297005113222


In [3]:
standard_deviation = np.sqrt(dispersion)

In [4]:
standard_deviation

0.03845425464771515

In [5]:
avg_data

選手名     打率        偏差          偏差^2
0        山田哲人  0.329  0.078005  6.084843e-03
1        川端慎吾  0.336  0.085005  7.225919e-03
2          雄平  0.270  0.019005  3.612054e-04
3        畠山和洋  0.268  0.017005  2.891838e-04
4        中村悠平  0.231 -0.019995  3.997838e-04
5        大引啓次  0.225 -0.025995  6.757189e-04
6        比屋根渉  0.230 -0.020995  4.407730e-04
7        上田剛史  0.263  0.012005  1.441298e-04
8      Ｍ・デニング  0.222 -0.028995  8.406865e-04
9        荒木貴裕  0.253  0.002005  4.021651e-06
10       田中浩康  0.201 -0.049995  2.499459e-03
11       今浪隆博  0.317  0.066005  4.356714e-03
12       森岡良介  0.188 -0.062995  3.968319e-03
13       三輪正義  0.233 -0.017995  3.238054e-04
14       坂本勇人  0.269  0.018005  3.241946e-04
15       長野久義  0.251  0.000005  2.921841e-11
16       亀井善行  0.272  0.021005  4.412271e-04
17      阿部慎之助  0.242 -0.008995  8.090273e-05
18       片岡治大  0.244 -0.006995  4.892435e-05
19       村田修一  0.236 -0.014995  2.248379e-04
20      立岡宗一郎  0.304  0.053005  2.809573e-03
21       井端弘和  0.234 -0.016995  2.888162e-04
22   Ｌ・アンダーソン  0.252  0.001005  1.010840e-06
23       小林誠司  0.226 -0.024995  6.247298e-04
24       高橋由伸  0.278  0.027005  7.292919e-04
25        橋本到  0.219 -0.031995  1.023654e-03
26       大田泰示  0.277  0.026005  6.762811e-04
27       相川亮二  0.313  0.062005  3.844670e-03
28       堂上剛裕  0.276  0.025005  6.252703e-04
29        鳥谷敬  0.281  0.030005  9.003244e-04
..        ...    ...       ...           ...
155       伊藤光  0.271  0.020005  4.002162e-04
156   Ｆ・カラバイヨ  0.252  0.001005  1.010840e-06
157    Ｅ・ヘルマン  0.267  0.016005  2.561730e-04
158      西野真弘  0.304  0.053005  2.809573e-03
159     小谷野栄一  0.295  0.044005  1.936476e-03
160    Ｔ・ブランコ  0.194 -0.056995  3.248384e-03
161      山崎勝己  0.204 -0.046995  2.208492e-03
162      縞田拓弥  0.226 -0.024995  6.247298e-04
163      宮﨑祐樹  0.252  0.001005  1.010840e-06
164       原拓也  0.170 -0.080995  6.560124e-03
165      川端崇義  0.239 -0.011995  1.438703e-04
166      平野恵一  0.262  0.011005  1.211189e-04
167      坂口智隆  0.262  0.011005  1.211189e-04
168      岩崎恭平  0.281  0.030005  9.003244e-04
169     松井稼頭央  0.256  0.005005  2.505408e-05
170    Ｗ・ペーニャ  0.268  0.017005  2.891838e-04
171      藤田一也  0.270  0.019005  3.612054e-04
172      後藤光尊  0.237 -0.013995  1.958487e-04
173       嶋基宏  0.219 -0.031995  1.023654e-03
174        銀次  0.301  0.050005  2.500541e-03
175   Ｚ・ウィーラー  0.255  0.004005  1.604327e-05
176       聖澤諒  0.252  0.001005  1.010840e-06
177   Ｇ・サンチェス  0.226 -0.024995  6.247298e-04
178      福田将儀  0.216 -0.034995  1.224622e-03
179      中川大志  0.235 -0.015995  2.558271e-04
180      西田哲朗  0.220 -0.030995  9.606649e-04
181      牧田明久  0.311  0.060005  3.600649e-03
182      岡島豪郎  0.168 -0.082995  6.888103e-03
183      伊志嶺忠  0.241 -0.009995  9.989192e-05
184      阿部俊人  0.196 -0.054995  3.024405e-03

[185 rows x 4 columns]

In [6]:
avg_data.mean()['打率']

0.25099459459459456

In [7]:
min_range = avg_data.mean()['打率'] - standard_deviation

In [8]:
max_range = avg_data.mean()['打率'] + standard_deviation

In [9]:
min_range

0.2125403399468794

In [10]:
max_range

0.28944884924230974

In [11]:
avg_data_2 = org_data[['選手名','打率']]

In [12]:
avg_data_2 = avg_data_2.astype({'打率': 'float64'})

In [13]:
avg_data_2.sort_values('打率', ascending=False, inplace=True)

In [16]:
avg_data_2

選手名     打率
0       柳田悠岐  0.363
1       秋山翔吾  0.359
2       川端慎吾  0.336
3       山田哲人  0.329
4       近藤健介  0.326
5       清田育宏  0.317
6       筒香嘉智  0.317
7       今浪隆博  0.317
8       相川亮二  0.313
9       牧田明久  0.311
10    Ｒ・ナニータ  0.308
11     立岡宗一郎  0.304
12      西野真弘  0.304
13        銀次  0.301
14       中村晃  0.300
15       荒波翔  0.298
16      和田一浩  0.298
17     小谷野栄一  0.295
18      藤井淳志  0.295
19      杉谷拳士  0.295
20      脇谷亮太  0.294
21      角中勝也  0.293
22      Ｈ・ルナ  0.292
23     Ｊ・ロペス  0.291
24     伊志嶺翔大  0.291
25      宮﨑敏郎  0.289
26      今江敏晃  0.287
27      松田宣浩  0.287
28       森友哉  0.287
29      淺間大基  0.285
..       ...    ...
155      嶋基宏  0.219
156     吉村裕基  0.217
157     福田将儀  0.216
158     江越大賀  0.214
159    炭谷銀仁朗  0.211
160  Ｊ・ハーミッダ  0.211
161     高橋周平  0.208
162     倉本寿彦  0.208
163     山崎勝己  0.204
164     大谷翔平  0.202
165     髙城俊人  0.202
166     田中浩康  0.201
167     阿部俊人  0.196
168     鬼﨑裕司  0.196
169     鶴岡慎也  0.195
170     大野奨太  0.194
171   Ｔ・ブランコ  0.194
172     新井良太  0.193
173     森岡良介  0.188
174     外崎修汰  0.186
175     桑原将志  0.184
176     杉山翔大  0.183
177    黒羽根利規  0.178
178     髙谷裕亮  0.175
179     藤井彰人  0.172
180     田村龍弘  0.170
181      原拓也  0.170
182     岡島豪郎  0.168
183     松井雅人  0.135
184      細川亨  0.116

[185 rows x 2 columns]

In [15]:
avg_data_2.reset_index(drop=True, inplace=True)

## 標準化得点
標準化得点 = (元のデータ - 平均値) ÷ 標準偏差

### 例
2015山田哲人の標準化得点

- (0.329 - 0.251) ÷ 0.038 = 2.03

## 偏差値
```
標準化得点 = (元のデータ - 平均値) ÷ 標準偏差 × 10 + 50
```

### 例
2015山田哲人の打率偏差値

- (0.329 - 0.251) ÷ 0.038 × 10 + 50 = 70.3

In [20]:
testes = (0.329 - avg_data.mean()['打率']) / standard_deviation * 10 + 50
print(testes)

70.28524700843222


## 非線型変換
GB/FBみたいな、比率のような間隔が等しくないものを比較する際に元の値を変換する手法
```
log10(元のデータ)
```

### 箱ひげ図
最小値と最大値、四分位偏差と中央値を一気に確認できるグラフ（メモ）

### グラフの最大値と最小値
大体集めたデータの平均値±2標準偏差の値をそれぞれ最大値、最小値とするといい（らしい）

# 母集団と標本

## 母集団と標本
```
母集団とは全てのデータの集まりを指し、標本とはそこからサンプリングした（集めた一部の）データのこと。
少ないデータ（標本）から、全体（母集団）を推測するために、標本誤差という、標本の値と母集団の値との誤差を意識する必要がある。
```

## 標本平均と母平均
```
全体のデータをいくつかのグループに分け（標本化）、そのグループ、要は標本内での平均をとったものが「標本平均」
そして、全グループの標本平均と全体のデータの平均（母平均）は一致する。このことを「標本平均は不偏性を持つ」という。
この性質があるために、標本平均の値はそのまま母集団の平均推測値として用いることができる。
```

## 不偏分散

- Σ((データ - 平均値)^2) / (データ数(N) - 1)

## 不偏分散を用いる理由
```
標本分散は、平均が母分散の値と一致しない（不偏性を持たない）ため、母分散の仮の値として用いることが出来ない。
そのため、不偏分散という値を算出して、それを母分散の仮の値とする。
```

In [2]:
org_data_4 = pd.read_csv('./statistics_data/4章サンプルデータ/Sheet1-表1.csv')

In [3]:
org_data_4

No    Group     チーム        登録名        生年月日   身長   体重
0      1   Group1      楽天       上園啓史   1984/6/30  182   84
1      2   Group1    ヤクルト       石山泰稚    1988/9/1  182   75
2      3   Group1    日本ハム       西川遥輝   1992/4/16  179   73
3      4   Group1    日本ハム       上沢直之    1994/2/6  187   88
4      5   Group1      阪神        鶴直人   1987/4/25  180   77
5      6   Group1    日本ハム       佐藤賢治   1988/8/26  181   87
6      7   Group1  ソフトバンク       明石健志    1986/1/9  175   68
7      8   Group1      西武       中崎雄太   1991/3/11  180   83
8      9   Group1      阪神      トラヴィス   1993/10/9  191   84
9     10   Group1   オリックス       齋藤俊雄  1983/12/23  180   85
10    11   Group1   オリックス       西野真弘    1990/8/2  167   68
11    12   Group1     ロッテ       大嶺翔太   1991/9/17  180   78
12    13   Group1    ヤクルト       中島彰吾   1992/6/15  178   72
13    14   Group1      楽天         銀次   1988/2/24  174   85
14    15   Group1      巨人       金城龍彦   1976/7/27  177   82
15    16   Group1      広島     Ｊ・グスマン   1984/6/14  185   91
16    17   Group1      阪神        岩本輝  1992/10/21  182   80
17    18   Group1    ヤクルト       奥村展征   1995/5/26  177   72
18    19   Group1    ヤクルト       西田明央   1992/4/28  178   83
19    20   Group1      巨人       澤村拓一    1988/4/3  183   98
20    21   Group1      巨人        青山誠   1991/11/1  185   85
21    22   Group1      楽天      八百板卓丸   1997/1/17  180   76
22    23   Group1   オリックス        谷佳知    1973/2/9  173   77
23    24   Group1     ロッテ       木村優太   1985/5/21  190   85
24    25   Group1      阪神       二神一人    1987/6/3  185   87
25    26   Group1      阪神       藤原正典   1988/1/14  182   81
26    27   Group1    日本ハム        石川亮   1995/7/20  180   83
27    28   Group1      巨人       長江翔太   1991/10/8  190   88
28    29   Group1     ロッテ     Ｃ・ハフマン   1985/4/29  183   98
29    30   Group1    日本ハム        武田久  1978/10/14  170   73
..   ...      ...     ...        ...         ...  ...  ...
860  861  Group10      巨人       久保裕也   1980/5/23  177   80
861  862  Group10   オリックス     Ｔ・ブランコ  1980/11/10  188  102
862  863  Group10      広島       菊池涼介   1990/3/11  171   69
863  864  Group10      広島       磯村嘉孝   1992/11/1  178   82
864  865  Group10  ソフトバンク       高谷裕亮  1981/11/13  179   86
865  866  Group10    ヤクルト       中元勇作  1988/12/12  178   73
866  867  Group10  ソフトバンク      長谷川勇也  1984/12/22  180   88
867  868  Group10   オリックス       若月健矢   1995/10/4  179   85
868  869  Group10      楽天       小斉祐輔   1983/4/30  178   88
869  870  Group10     ロッテ       今江敏晃   1983/8/26  180   89
870  871  Group10    ヤクルト     松元ユウイチ  1980/12/18  178   86
871  872  Group10      楽天       安樂智大   1996/11/4  186   87
872  873  Group10    DeNA  Ｌ・グリエルJr.  1993/10/19  192   83
873  874  Group10  ソフトバンク       山田大樹   1988/7/30  189   95
874  875  Group10      巨人       隠善智也    1984/6/5  174   77
875  876  Group10      広島       藤井皓哉   1996/7/29  182   83
876  877  Group10      巨人       田原啓吾   1994/6/12  182   86
877  878  Group10      広島       野間峻祥   1993/1/28  180   80
878  879  Group10      広島       中村亘佑    1991/4/2  181   83
879  880  Group10      広島       鈴木誠也   1994/8/18  181   83
880  881  Group10   オリックス        小松聖  1981/10/29  180   80
881  882  Group10      楽天       釜田佳直  1993/10/26  177   78
882  883  Group10    DeNA       赤堀大智   1987/4/13  188   90
883  884  Group10    日本ハム       岸里亮佑    1995/4/3  180   75
884  885  Group10      中日       石垣幸大    1996/8/8  180   76
885  886  Group10      中日       福田永将   1988/7/23  181   80
886  887  Group10      中日       藤井淳志   1981/5/20  182   81
887  888  Group10      西武       山田遥楓   1996/9/30  179   78
888  889  Group10      中日       浜田智博   1992/10/1  183   75
889  890  Group10      中日       友永翔太    1991/4/1  170   74

[890 rows x 7 columns]

In [4]:
grouped = org_data_4.groupby('Group')

In [23]:
grouped.mean().drop(['No','体重'], axis=1).mean()

身長    180.780899
dtype: float64

In [18]:
grouped.agg({'No': 'count', '身長': 'std', '体重': 'mean'})

No        身長         体重
Group                           
Group1   89  5.678144  82.078652
Group10  89  5.005181  82.314607
Group2   89  4.248325  83.269663
Group3   89  6.428021  83.191011
Group4   89  6.483538  83.988764
Group5   89  4.901715  83.505618
Group6   89  5.498166  83.550562
Group7   89  5.688007  83.876404
Group8   89  5.867167  81.674157
Group9   89  4.362149  81.685393

In [16]:
grouped.describe().drop(['No','体重'], axis=1)['身長']

count        mean       std    min    25%    50%    75%    max
Group                                                                  
Group1    89.0  180.606742  5.678144  167.0  177.0  180.0  183.0  198.0
Group10   89.0  180.247191  5.005181  170.0  177.0  180.0  183.0  195.0
Group2    89.0  181.494382  4.248325  171.0  178.0  181.0  184.0  193.0
Group3    89.0  181.460674  6.428021  168.0  178.0  181.0  185.0  205.0
Group4    89.0  180.685393  6.483538  163.0  177.0  180.0  185.0  198.0
Group5    89.0  181.202247  4.901715  168.0  177.0  181.0  184.0  198.0
Group6    89.0  180.516854  5.498166  170.0  177.0  180.0  183.0  198.0
Group7    89.0  181.146067  5.688007  166.0  178.0  180.0  185.0  201.0
Group8    89.0  180.089888  5.867167  163.0  176.0  180.0  183.0  197.0
Group9    89.0  180.359551  4.362149  171.0  177.0  180.0  184.0  191.0

In [ ]:
grouped.agg(lambda x: )

In [20]:
grouped.var().drop(['No','体重'], axis=1)

身長
Group             
Group1   32.241318
Group10  25.051839
Group2   18.048264
Group3   41.319459
Group4   42.036261
Group5   24.026813
Group6   30.229826
Group7   32.353422
Group8   34.423647
Group9   19.028345

In [21]:
grouped.var(ddof=False).drop(['No','体重'], axis=1)

身長
Group             
Group1   31.879056
Group10  24.770357
Group2   17.845474
Group3   40.855195
Group4   41.563944
Group5   23.756849
Group6   29.890165
Group7   31.989900
Group8   34.036864
Group9   18.814544

## 信頼区間とは

- 母平均が含まれると推定されるデータの幅で、母平均の値が含まれる範囲の確率を示す形で表される。

`95%信頼区間とは、信頼区間の範囲内に、母平均が９5%の確率で含まれることを示す。`

## 信頼区間を求める式
$$
    信頼区間 = 標本平均 \pm t\sqrt{\frac{不偏分散}{データ数(N)}}
$$

## 信頼区間を求める式（比率の場合）
$$
    信頼区間 = 標本比率(p) \pm Z\sqrt{\frac{p(1-p)}{データ数(N)}}
$$

`※Zは標準正規分布の値を当てる`

In [26]:
grouped.agg(['count', 'mean', 'var']).drop(['No','体重'], axis=1)['身長']

count        mean        var
Group                                
Group1      89  180.606742  32.241318
Group10     89  180.247191  25.051839
Group2      89  181.494382  18.048264
Group3      89  181.460674  41.319459
Group4      89  180.685393  42.036261
Group5      89  181.202247  24.026813
Group6      89  180.516854  30.229826
Group7      89  181.146067  32.353422
Group8      89  180.089888  34.423647
Group9      89  180.359551  19.028345

In [29]:
# t値はExcelの値を参考に一旦1.99で設定
confidence_interval = grouped.agg(['count', 'mean', 'var']).drop(['No','体重'], axis=1)['身長']
confidence_interval['ci_min'] = confidence_interval['mean'] - 1.99 * np.sqrt(confidence_interval['var'] / confidence_interval['count'])
confidence_interval['ci_max'] = confidence_interval['mean'] + 1.99 * np.sqrt(confidence_interval['var'] / confidence_interval['count'])

In [30]:
confidence_interval

count        mean        var      ci_min      ci_max
Group                                                        
Group1      89  180.606742  32.241318  179.408996  181.804487
Group10     89  180.247191  25.051839  179.191400  181.302982
Group2      89  181.494382  18.048264  180.598242  182.390522
Group3      89  181.460674  41.319459  180.104750  182.816598
Group4      89  180.685393  42.036261  179.317759  182.053028
Group5      89  181.202247  24.026813  180.168281  182.236213
Group6      89  180.516854  30.229826  179.357073  181.676635
Group7      89  181.146067  32.353422  179.946242  182.345893
Group8      89  180.089888  34.423647  178.852270  181.327505
Group9      89  180.359551  19.028345  179.439401  181.279700

## 信頼区間で打率予想！
$$
    打率信頼区間 = 去年の打率(p) \pm 1.96\sqrt{\frac{p(1-p)}{打数}}
$$

In [31]:
def avg_estimation(avg, at_bat):
    min_avg = avg - 1.96 * np.sqrt(avg * (1 - avg) / at_bat)
    max_avg = avg + 1.96 * np.sqrt(avg * (1 - avg) / at_bat)
    
    print('2020予想最低打率： {}'.format(min_avg))
    print('2020予想最高打率： {}'.format(max_avg))

In [32]:
avg_estimation(0.2711, 520)

2020予想最低打率： 0.23289210308004177
2020予想最高打率： 0.30930789691995825
